cloud/amazon/ec2_ami.py

```diff
@@ -392,10 +540,18 @@ def main():
 
         # Changed is always set to true when provisioning new AMI
         if not module.params.get('instance_id'):
-            module.fail_json(msg='instance_id parameter is required for new image')
-        if not module.params.get('name'):
-            module.fail_json(msg='name parameter is required for new image')
-        create_image(module, ec2)
+            if not module.params.get('image_location') and not module.params.get('device_mapping'):
+                module.fail_json(msg='image_location (register from s3) or device_mapping (register from ebs snapshot) parameter is required for new image')
+
+            register_image(module, ec2)
+        else:
+            if module.params.get('image_location'):
+                module.fail_json(msg='instance_id and image_location parameters are mutually exclusive')
+
+            if not module.params.get('name'):
+                module.fail_json(msg='name parameter is required for new image')
+
+            create_image(module, ec2)
 
 
 # import module snippets
```

Comment:
Sorry, one last thing and I'll leave you alone :)
Can you improve this:
   main()

to become this:
    if __name__ == '__main__':
        main()